### PR TITLE
[mypy] fix directory "dynamic_programming"

### PR DESCRIPTION
**Pull request by algobytewise**
### **Describe your change:**
Related Issue: #4052

* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [ ] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [ ] All filenames are in all lowercase characters with no spaces or dashes.
* [ ] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [ ] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by dhruvmanila** [comment]
We are not only correcting the `mypy` errors, but actually providing type hints where they are missing like in: https://github.com/TheAlgorithms/Python/blob/master/dynamic_programming/bitmask.py#L15

I was already working on `dynamic_programming` directory and have fixed around 13 files. Our goal is not to satisfy `mypy` but actually get the entire codebase statically typed.

File: dynamic_programming/fractional_knapsack_2.py [hunk 1]
```diff
@@ -49,10 +49,12 @@ def fractional_knapsack(
 
 if __name__ == "__main__":
     n = int(input("Enter number of items: "))
-    value = input(f"Enter the values of the {n} item(s) in order: ").split()
-    value = [int(v) for v in value]
-    weight = input(f"Enter the positive weights of the {n} item(s) in order: ".split())
-    weight = [int(w) for w in weight]
+    valueStrings = input(f"Enter the values of the {n} item(s) in order: ").split()
```
Comment (dhruvmanila) [inline dynamic_programming/fractional_knapsack_2.py:52]: Python follows the `snake_case` naming convention.

File: mypy.ini [hunk 2]
```diff
@@ -2,4 +2,4 @@
 ignore_missing_imports = True
 
 ; FIXME: #4052 fix mypy errors in the exclude directories and remove them below
-exclude = (data_structures|dynamic_programming|graphs|maths|matrix|other|project_euler|searches|strings*)/$
+exclude = (data_structures|graphs|maths|matrix|other|project_euler|searches|strings*)/$
```
Comment (dhruvmanila) [inline mypy.ini:5]: Please remove this change as the directory is not completely typed.

**Comment by algobytewise** [comment]
With getting "the entire codebase statically typed" I assume you mean all the inputs & outputs of functions and not truly everything.

File: mypy.ini (hunk 2 above)
Comment (algobytewise) [reply mypy.ini:5]: If we want to make sure that there are no untyped functions, we could use the additional parameter `--disallow-incomplete-defs`, see https://mypy.readthedocs.io/en/stable/command_line.html#untyped-definitions-and-calls.

**Comment by dhruvmanila** [comment]
I do mean everything which `mypy` cannot infer itself. Some variables will be difficult for `mypy` to infer the type of for which we should provide the type. Common examples include the container types like `list`, `dict`, etc.

File: mypy.ini (hunk 2 above)
Comment (dhruvmanila) [reply mypy.ini:5]: Thank you looking into it. We are going to make `mypy` stricter, but not now.

**Comment by algobytewise** [comment]
Ah I see, that's good to know.

**Comment by algobytewise** [comment]
Could someone have a look at this contribution? It fixes all the errors (`mypy --ignore-missing-imports .`) but it doesn't fix all the missing type annotations (`mypy --ignore-missing-imports --disallow-untyped-defs .`). In the meantime there has been another pull request for fixing this directory: #4332. Not sure which one is a better candidate for the merge.

**Pull request closed without merging**